### ISSUE TITLE
Bug when using scan + grad + vmap + custom_vjp

### ISSUE THREAD

**Issue by EddieCunningham** [google/jax#4173]
I found a bug that appears when using the combination of lax.scan, jax.grad, jax.vmap and jax.custom_vjp.  Here is how to reproduce the bug:

```python

import jax
import itertools
import jax.numpy as jnp
from functools import partial

# Create the input data.
x = jnp.ones((10, 3))

# Create the custom function
@jax.custom_vjp
def custom_fun(x):
    return x.sum()

def forward(x):
    return x.sum(), (jnp.ones_like(x),)

def backward(res, g):
    return g*res[0],

custom_fun.defvjp(forward, backward)

# This function fails inside of a scan loop!
def train_fun(x):
    def summed_fun(x):
        return jax.vmap(custom_fun)(x).sum()
    return jax.grad(summed_fun)(x)

def scan_body(carry, inputs):
    x = carry
    return carry, train_fun(x)

scan_range = jnp.arange(4)
jax.lax.scan(scan_body, x, scan_range) # Fails!

```
Here is the error message
```
Traceback (most recent call last):
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/eddie/NuX/bug_simple.py", line 33, in <module>
    jax.lax.scan(scan_body, x, scan_range) # Fails!
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/lax/lax_control_flow.py", line 1232, in scan
    jaxpr, consts, out_tree = _initial_style_jaxpr(f, in_tree, carry_avals + x_avals)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/lax/lax_control_flow.py", line 74, in _initial_style_jaxpr
    fun, in_tree, in_avals)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/lax/lax_control_flow.py", line 68, in _initial_style_untyped_jaxpr
    wrapped_fun, in_pvals, instantiate=True, stage_out=False)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/interpreters/partial_eval.py", line 423, in trace_to_jaxpr
    jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/linear_util.py", line 150, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/eddie/NuX/bug_simple.py", line 30, in scan_body
    return carry, train_fun(x)
  File "/home/eddie/NuX/bug_simple.py", line 26, in train_fun
    return jax.grad(summed_fun)(x)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/api.py", line 429, in grad_f
    _, g = value_and_grad_f(*args, **kwargs)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/api.py", line 487, in value_and_grad_f
    ans, vjp_py = _vjp(f_partial, *dyn_args)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/api.py", line 1516, in _vjp
    out_primal, out_vjp = ad.vjp(flat_fun, primals_flat)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/interpreters/ad.py", line 110, in vjp
    out_primals, pvals, jaxpr, consts = linearize(traceable, *primals)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/interpreters/ad.py", line 97, in linearize
    jaxpr, out_pvals, consts = pe.trace_to_jaxpr(jvpfun_flat, in_pvals)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/interpreters/partial_eval.py", line 423, in trace_to_jaxpr
    jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/linear_util.py", line 150, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/eddie/NuX/bug_simple.py", line 25, in summed_fun
    return jax.vmap(custom_fun)(x).sum()
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/api.py", line 889, in batched_fun
    lambda: flatten_axes("vmap out_axes", out_tree(),
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/interpreters/batching.py", line 35, in batch
    return batched_fun.call_wrapped(*in_vals)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/linear_util.py", line 150, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/custom_derivatives.py", line 474, in __call__
    *args_flat, out_trees=out_trees)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/custom_derivatives.py", line 541, in custom_vjp_call_jaxpr
    out_trees=out_trees)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/core.py", line 279, in bind
    out_tracer = top_trace.process_primitive(self, tracers, kwargs)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/interpreters/batching.py", line 137, in process_primitive
    return map(partial(BatchTracer, self), val_out, dim_out)
  File "/home/eddie/anaconda3/envs/flow/lib/python3.7/site-packages/jax/util.py", line 33, in safe_map
    assert len(arg) == n, 'length mismatch: {}'.format(list(map(len, args)))
AssertionError: length mismatch: [1, 2]
```

It seems like this is the only combination of scan, grad, vmap and custom_vjp that fails.  I tested each combination in a test script (sorry for the block of code, but it doesn't look like I can attach python files to issues!)

```python
import jax
import itertools
import jax.numpy as jnp
from functools import partial

# Create the input data.
x = jnp.ones((10, 3))

####################################
# Baseline function.

def fun(x):
    return x.sum()

####################################
# Custom vjp function.  Expect same functionality as fun.

@jax.custom_vjp
def custom_fun(x):
    return x.sum()

def forward(x):
    return x.sum(), (jnp.ones_like(x),)

def backward(res, g):
    return g*res[0],

custom_fun.defvjp(forward, backward)

####################################
# Functions that will be applied to fun or custom_fun
# with or without vmap.

def make_train_fun(custom=False, vmap=False, grad=False):

    def train_fun(x):

        # Use the custom_vjp function or baseline.
        after_custom = custom_fun if custom else fun

        # Should we vmap the function?
        after_vmap = jax.vmap(after_custom) if vmap else after_custom

        # We need to sum the result to get a scalr so that grad works.
        after_sum = lambda x: after_vmap(x).sum()

        # Should we grad the function?
        after_grad = jax.grad(after_sum) if grad else after_sum
        return after_grad(x)

    # Print out the equivalent function call for clarity
    custom_str = 'custom_fun' if custom else 'fun'
    vmap_str   = 'jax.vmap(%s)'%custom_str if vmap else custom_str
    grad_str   = 'jax.grad(summed_fun)' if grad else 'summed_fun'
    fun_string = ('def train_fun(x):\n'
                 '    def summed_fun(x):\n'
                 '        return %s(x).sum()\n'
                 '    return %s(x)\n'
                 '\n'
                 'train_fun(x)\n')%(vmap_str, grad_str)

    scan_str = ('def scan_body(carry, inputs):\n'
                '    x = carry\n'
                '    return carry, train_fun(x)\n'
                '\n'
                'jax.lax.scan(scan_body, x, scan_range)\n')

    return train_fun, fun_string + '\n' + scan_str

####################################
# We will be scanning

def scan_body(f, carry, inputs):
    x = carry
    return carry, f(x)

scan_range = jnp.arange(4)
scan_apply = lambda f: jax.lax.scan(partial(scan_body, f), x, scan_range)

####################################

for custom, vmap, grad in itertools.product(*([[False, True]]*3)):
    print('\n# -------------------------\n')

    train_fun, code = make_train_fun(custom=custom, vmap=vmap, grad=grad)
    print('# Code:\n')
    print(code)

    try:
        train_fun(x)
        print('# Passed function call!')
    except:
        print('# FAILED function call with error:\n\n')
        assert 0

    try:
        scan_apply(train_fun)
        print('# Passed scan call!')
    except:
        print('# FAILED scan call with error:\n\n')
        assert 0

```

I've attached the output of this script to a text file.
[bug_output.txt](https://github.com/google/jax/files/5143586/bug_output.txt)

Any ideas on what might be going on or a workaround that I can use?  Thanks!

**Comment by j-towns** [comment]
Just trying to reproduce this on my laptop. Your first code snipped gives
```text
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-1-2471aa99d9dc> in <module>
     30     return carry, train_fun(x)
     31
---> 32 jax.lax.scan(scan_body, x, scan_range) # Fails!

NameError: name 'scan_range' is not defined
```
What shape is `scan_range` supposed to have?

**Comment by EddieCunningham** [comment]
Sorry about that!  I used
```python
scan_range = jnp.arange(4)
```
I just updated the original comment.

**Comment by mattjj** [comment]
This was fixed by #4008 !
